File: adapters/sqlite/hosts/macos/Sources/EngawaSQLite/SqliteAdapter.swift
import Foundation
import EngawaKit
import SQLite3

// The `sqlite` reference adapter (adapters/sqlite/spec.md). Request-driven, local, durable —
// the first consumer of the adapter SDK from outside the built-in set. Text/number values
// over the wire; blobs are out of the v1 text scope.
private let SQLITE_TRANSIENT = unsafeBitCast(-1, to: sqlite3_destructor_type.self)

public final class SqliteAdapter: Adapter, @unchecked Sendable {
    public let namespace = "sqlite"

    private let lock = NSLock()   // guards dbs/nextHandle; @unchecked Sendable rests on it
    private var dbs: [Int: OpaquePointer] = [:]
    private var nextHandle = 1

    public init() {}

    public func handle(_ cmd: String, _ args: JSONValue) async throws -> JSONValue {
        try perform(cmd, args)
    }

    // Synchronous so NSLock is legal (locks are unavailable from async contexts in Swift 6).
    // SQLite handles are not concurrency-safe, so the whole operation is serialized.
    private func perform(_ cmd: String, _ args: JSONValue) throws -> JSONValue {
        lock.lock(); defer { lock.unlock() }
        switch cmd {
        case "open":    return try open(args)
        case "execute": return try run(args, query: false)
        case "query":   return try run(args, query: true)
        case "close":   return try close(args)
        default:        throw AdapterError("ENOSYS", "unknown command: sqlite.\(cmd)")
        }
    }

    private func open(_ args: JSONValue) throws -> JSONValue {
        guard let path = args.objectValue?["path"]?.stringValue, !path.isEmpty else {
            throw AdapterError("EINVAL", "path required")
        }
        var db: OpaquePointer?
        if sqlite3_open(path, &db) != SQLITE_OK {
            let msg = db.map { String(cString: sqlite3_errmsg($0)) } ?? "open failed"
            if let db = db { sqlite3_close(db) }
            throw AdapterError("ESQLITE", msg)
        }
        let handle = nextHandle
        nextHandle += 1
        dbs[handle] = db
        return .object(["db": .number(Double(handle))])
    }

    private func run(_ args: JSONValue, query: Bool) throws -> JSONValue {
        let obj = args.objectValue ?? [:]
        guard let handle = obj["db"]?.numberValue else { throw AdapterError("EINVAL", "db required") }
        guard let sql = obj["sql"]?.stringValue else { throw AdapterError("EINVAL", "sql required") }
        guard let db = dbs[Int(handle)] else { throw AdapterError("EBADF", "unknown db handle: \(Int(handle))") }

        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK, let stmt = stmt else {
            throw AdapterError("ESQLITE", String(cString: sqlite3_errmsg(db)))
        }
        defer { sqlite3_finalize(stmt) }

        try bind(obj["params"]?.arrayValue ?? [], to: stmt)

        if query {
            var rows: [JSONValue] = []
            let cols = Int(sqlite3_column_count(stmt))
            while sqlite3_step(stmt) == SQLITE_ROW {
                var row: [String: JSONValue] = [:]
                for c in 0..<cols {
                    let name = String(cString: sqlite3_column_name(stmt, Int32(c)))
                    row[name] = columnValue(stmt, Int32(c))
                }
                rows.append(.object(row))
            }
            return .object(["rows": .array(rows)])
        } else {
            let rc = sqlite3_step(stmt)
            guard rc == SQLITE_DONE || rc == SQLITE_ROW else {
                throw AdapterError("ESQLITE", String(cString: sqlite3_errmsg(db)))
            }
            return .object([
                "changes": .number(Double(sqlite3_changes(db))),
                "lastInsertRowid": .number(Double(sqlite3_last_insert_rowid(db))),
            ])
        }
    }

    private func close(_ args: JSONValue) throws -> JSONValue {
        guard let handle = args.objectValue?["db"]?.numberValue else { throw AdapterError("EINVAL", "db required") }
        guard let db = dbs.removeValue(forKey: Int(handle)) else { throw AdapterError("EBADF", "unknown db handle") }
        sqlite3_close(db)
        return .null
    }

    private func bind(_ params: [JSONValue], to stmt: OpaquePointer) throws {
        for (i, p) in params.enumerated() {
            let idx = Int32(i + 1)
            switch p {
            case .null:
                sqlite3_bind_null(stmt, idx)
            case .bool(let b):
                sqlite3_bind_int64(stmt, idx, b ? 1 : 0)
            case .number(let d):
                if d.rounded() == d && abs(d) < 9.007e15 { sqlite3_bind_int64(stmt, idx, Int64(d)) }
                else { sqlite3_bind_double(stmt, idx, d) }
            case .string(let s):
                sqlite3_bind_text(stmt, idx, s, -1, SQLITE_TRANSIENT)
            default:
                throw AdapterError("EINVAL", "unsupported parameter type at index \(i)")
            }
        }
    }

    private func columnValue(_ stmt: OpaquePointer, _ c: Int32) -> JSONValue {
        switch sqlite3_column_type(stmt, c) {
        case SQLITE_INTEGER: return .number(Double(sqlite3_column_int64(stmt, c)))
        case SQLITE_FLOAT:   return .number(sqlite3_column_double(stmt, c))
        case SQLITE_TEXT:    return .string(String(cString: sqlite3_column_text(stmt, c)))
        default:             return .null   // NULL, and BLOB (out of v1 text scope)
        }
    }
}

File: adapters/sqlite/hosts/linux/SqliteAdapter.cpp
// The `sqlite` namespace on Windows (adapters/sqlite/spec.md) — durable local storage that survives
// independently of WebView-managed storage (contract §10). A reference adapter: request-driven,
// local, durable. Booleans bind as 0/1; blobs are out of the v1 text scope (§5a covers binary). Each
// open is a distinct connection closed on `close`, so the file handle is released (a later fs.remove
// on Windows would otherwise hit a sharing violation).
#include <cmath>
#include <unordered_map>

#include "adapters/Adapters.hpp"
#include "sqlite3.h"

namespace {

Json columnValue(sqlite3_stmt* stmt, int i) {
    switch (sqlite3_column_type(stmt, i)) {
        case SQLITE_INTEGER:
            return Json(static_cast<long long>(sqlite3_column_int64(stmt, i)));
        case SQLITE_FLOAT:
            return Json(sqlite3_column_double(stmt, i));
        case SQLITE_NULL:
            return Json(nullptr);
        case SQLITE_TEXT:
        case SQLITE_BLOB:
        default: {
            const char* txt = reinterpret_cast<const char*>(sqlite3_column_text(stmt, i));
            int n = sqlite3_column_bytes(stmt, i);
            return Json(std::string(txt ? txt : "", n));
        }
    }
}

void bindParam(sqlite3_stmt* stmt, int idx, const Json& v) {
    if (v.is_null()) sqlite3_bind_null(stmt, idx);
    else if (v.is_boolean()) sqlite3_bind_int64(stmt, idx, v.get<bool>() ? 1 : 0);  // SQLite has no bool
    else if (v.is_string()) {
        std::string s = v.get<std::string>();
        sqlite3_bind_text(stmt, idx, s.c_str(), (int)s.size(), SQLITE_TRANSIENT);
    } else if (v.is_number_integer() || v.is_number_unsigned()) {
        sqlite3_bind_int64(stmt, idx, v.get<long long>());
    } else if (v.is_number_float()) {
        double d = v.get<double>();
        if (d == std::floor(d) && std::isfinite(d) && std::fabs(d) < 9.2e18)
            sqlite3_bind_int64(stmt, idx, static_cast<long long>(d));
        else
            sqlite3_bind_double(stmt, idx, d);
    } else {
        std::string s = v.dump();
        sqlite3_bind_text(stmt, idx, s.c_str(), (int)s.size(), SQLITE_TRANSIENT);
    }
}

class SqliteAdapter : public IAdapter {
public:
    ~SqliteAdapter() override {
        for (auto& [h, db] : dbs_) sqlite3_close(db);
    }

    std::string ns() const override { return "sqlite"; }

    Json handle(const std::string& command, const Json& args) override {
        if (command == "open") return open(ja::reqString(args, "path"));
        if (command == "execute") return execute(args);
        if (command == "query") return query(args);
        if (command == "close") return close(args);
        throw EngawaError::nosys("sqlite." + command);
    }

private:
    Json open(const std::string& path) {
        sqlite3* db = nullptr;
        int rc = sqlite3_open_v2(path.c_str(), &db, SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE, nullptr);
        if (rc != SQLITE_OK) {
            std::string msg = db ? sqlite3_errmsg(db) : "cannot open database";
            if (db) sqlite3_close(db);
            throw EngawaError("ESQLITE", msg);
        }
        int handle = next_++;
        dbs_[handle] = db;
        return Json{{"db", handle}};
    }

    // Prepare `sql` on the connection and bind the positional `?` params (spec: params is a positional
    // array — SQLite binds `?` by 1-based index natively, so no placeholder rewriting is needed).
    sqlite3_stmt* prepare(sqlite3* db, const Json& args) {
        std::string sql = ja::reqString(args, "sql");
        sqlite3_stmt* stmt = nullptr;
        if (sqlite3_prepare_v2(db, sql.c_str(), -1, &stmt, nullptr) != SQLITE_OK)
            throw EngawaError("ESQLITE", sqlite3_errmsg(db));
        if (args.is_object() && args.contains("params") && args["params"].is_array()) {
            const Json& p = args["params"];
            for (int i = 0; i < (int)p.size(); i++) bindParam(stmt, i + 1, p[i]);
        }
        return stmt;
    }

    Json execute(const Json& args) {
        sqlite3* db = handle(args);
        sqlite3_stmt* stmt = prepare(db, args);
        int rc;
        while ((rc = sqlite3_step(stmt)) == SQLITE_ROW) { /* drain any rows */ }
        if (rc != SQLITE_DONE) {
            std::string msg = sqlite3_errmsg(db);
            sqlite3_finalize(stmt);
            throw EngawaError("ESQLITE", msg);
        }
        sqlite3_finalize(stmt);
        return Json{{"changes", sqlite3_changes(db)},
                    {"lastInsertRowid", static_cast<long long>(sqlite3_last_insert_rowid(db))}};
    }

    Json query(const Json& args) {
        sqlite3* db = handle(args);
        sqlite3_stmt* stmt = prepare(db, args);
        Json rows = Json::array();
        int rc;
        while ((rc = sqlite3_step(stmt)) == SQLITE_ROW) {
            Json row = Json::object();
            int cols = sqlite3_column_count(stmt);
            for (int i = 0; i < cols; i++)
                row[sqlite3_column_name(stmt, i)] = columnValue(stmt, i);
            rows.push_back(std::move(row));
        }
        if (rc != SQLITE_DONE) {
            std::string msg = sqlite3_errmsg(db);
            sqlite3_finalize(stmt);
            throw EngawaError("ESQLITE", msg);
        }
        sqlite3_finalize(stmt);
        return Json{{"rows", rows}};
    }

    Json close(const Json& args) {
        int h = ja::reqInt(args, "db");
        auto it = dbs_.find(h);
        if (it == dbs_.end()) throw EngawaError("EBADF", "unknown db handle");
        sqlite3_close(it->second);
        dbs_.erase(it);
        return Json(nullptr);
    }

    sqlite3* handle(const Json& args) {
        int h = ja::reqInt(args, "db");
        auto it = dbs_.find(h);
        if (it != dbs_.end()) return it->second;
        throw EngawaError("EBADF", "unknown db handle");
    }

    std::unordered_map<int, sqlite3*> dbs_;
    int next_ = 1;
};

}  // namespace

std::unique_ptr<IAdapter> makeSqliteAdapter() { return std::make_unique<SqliteAdapter>(); }
